### ISSUE TITLE
Test if reporting services report is runnable

### ISSUE THREAD

**Issue by FuegoArtificial** [Seddryck/NBi#201]
Dear developers,

is there an aproach ready to just test a reporting services report for availability and if it's runnable? 
Testing a dataset is possible as far as I know, but it could be that some other dataset within the report fails while the one you test is running. Having a more general approach would be nice.
However, I see that testing all datasets in a report at once would achieve the same result.

I'm just asking, this is no feature request from my point of view for now.

P.S. Thanks for the frequent updates! I'm closely following the development and using nbi intensively.

Edit: Perhaps it would be achieved when issue #40 is implemented in the future?

Cheers,
Tilo

**Comment by Seddryck** [comment]
Hi @FuegoArtificial 

It depends by what you mean by a "report is runnable".

If I was requested to test a report, I'd like to test following steps:
1. The parameters are the expected parameters (None is missing, none is unexpected, correct types).
2. The report contains/links the expected datasets/linked-reports.
3. Each dataset used by my report behaves correctly (result returned is correct and performance)
4. My report correctly handles the datasets (presentation)

At the moment step 1, 2 and 4 are not covered by features. For step 3, you've [this feature](http://www.nbi.io/docs/query-report/) but it could be improved with #190 

My understanding is that you want to test the whole report in one test where I prefer to have many tests but when one of them is failing I know excalty what's wrong.

**Comment by Seddryck** [comment]
I should add that I'm working on [another project for the deployment of SSRS projects](https://github.com/Seddryck/RsPackage) where I need to learn a lot about internals of this tool and it will be easier for me to provide good solutions.

**Comment by FuegoArtificial** [comment]
You are right, I'm trying to test from a broad perspective which can be sufficient in some cases. For example if data structure changed, failing parameter or main data sets will be visible immediately. Maybe without instantly knowing which one, I know where to check. Especially, if I don't test for each parameter dataset. 
If getting results from the report in general it would point out that the data structure in the data source is still valid. 

For case 1 I built a template/solution to cover it. Although, it's not featured by nbi directly. Would you like to see that? 

I'll check your other project soon and might be able to support you with ideas while working with ssrs since 2004. Thanks for the tip! :-)

**Comment by Seddryck** [comment]
Feel free to share your solution for parameters will surely save me some time.

**Comment by FuegoArtificial** [comment]
Alright @Seddryck, I'm courious, what you think.

Workflow:
- Deploy all reports to the report server into some (sub-)folder tree
- Set all default parameters as required
- Read the default parameters from the reporting services database with the query included in the system-under-test tag without the constraint towards one report
- persist the values in an extra table (here `[dbo].[ReportDefaultParameters]` within the Nbi-Database)
- run the GenBIL-script combined with the template to get a test for each report
- run the test suite

template:

```
<test name="Check default parameters of report '$ReportName$'">
    $if(IgnoreReason)$<ignore>$IgnoreReason$</ignore>$endif$
    <description>Check default parameters of report '$ReportName$' of folder 'myFolder'</description>
    <edition author="$username$" created="$now$"/>
    <category>Default parameters</category>
    <system-under-test>
        <execution> 
            <query connectionString="@ReportServer">
                <![CDATA[
                DECLARE @PathLike AS VARCHAR(Max)
SET @PathLike = '/%' -- set subpath if applicable

SELECT
    ReportName,
    ParameterName,
    Prompt,
    v.value ('.', 'VARCHAR(Max)') AS Value
FROM
(
SELECT
    Params.p.value('Name[1]','varchar(Max)') AS ParameterName,
    Params.p.value('Prompt[1]','varchar(Max)') AS ParameterLabel,
    DefaultValueXML = Params.p.query('DefaultValues/Value'),
    Params.p.value ('DataType[1]', 'VARCHAR(Max)') AS DataType,
    Params.p.value ('AllowBlank[1]', 'VARCHAR(50)') AS AllowBlank,
    Params.p.value ('Prompt[1]', 'VARCHAR(Max)') AS Prompt,
    Params.p.value ('Hidden[1]', 'VARCHAR(Max)') AS Hidden,
    tmp.Name AS ReportName,
    tmp.Path AS ReportPath,
    CASE tmp.Type 
        WHEN 2 THEN 'Report'
        WHEN 4 THEN 'Linked Report'
    END AS ElementType

FROM
  (SELECT 
        Name
        , Path
        , ReportParams = CONVERT(XML,Parameter)
        , Type
    FROM dbo.Catalog 
    WHERE 
    (Type = 4 -- linked Report
    Or
    Type = 2) -- Report
    AND
    Path like @PathLike) tmp
    CROSS APPLY ReportParams.nodes('(Parameters/Parameter)') AS Params(p)
WHERE NOT Params.p.value('Name[1]','varchar(Max)') IN ('ReportPath','ReportNumber')) tmp2
CROSS APPLY DefaultValueXML.nodes('Value') value(v)
WHERE ReportName = '$ReportName$'
                ]]>
            </query>
        </execution>    
    </system-under-test>
    <assert>
        <equalTo>
            <column  index="0" role="key" type="text"/>
            <column  index="1" role="key" type="text"/>
            <column  index="2" role="key" type="text"/>
            <column  index="3" role="key" type="text"/>
            <query connectionString="@NBi">
                <![CDATA[
                SELECT [ReportName]
      ,[ParameterName]
      ,[Prompt]
      ,[Value]
  FROM [dbo].[ReportDefaultParameters]
WHERE ReportName = '$ReportName$'
                ]]>
            </query>
        </equalTo>
    </assert>
</test>
```

GenBIL-Script

```
setting reference 'NBi' connectionString 'Data Source=.;Initial Catalog=NBI;Integrated Security=SSPI';
setting reference 'ReportServer' connectionString 'Data Source=.;Initial Catalog=ReportServer;Integrated Security=SSPI';

case load query{
    /*CheckReportDefaultParameters*/
    /*Checks whether the fixed default values for report parameters remain the same on the server.*/

    SELECT DISTINCT   ReportName, IgnoreReason    FROM dbo.ReportDefaultParameters
}
on 'Data Source=.;Initial Catalog=NBI;Integrated Security=SSPI';

template load file 'D:\NBI\Templates\\CheckReportDefaultParameters.xml'; 
suite generate;

suite save 'D:\NBI\Tests\DefaultParameters.nbits';
```

Hope that helps to solve the requirement as long as there is no feature directly supporting it.
If you have questions, let me know.

Cheers,
Tilo

**Comment by Seddryck** [comment]
Useful code that I'll sure use when working on parameters. Thx for sharing.

I created a few tickets for features that NBi will support:

#203, #204, #205

**Comment by Seddryck** [comment]
Are you sure for the query and the property "Hidden"? Personaly I always receive a null (SQL Server 2016)

**Comment by FuegoArtificial** [comment]
I just checked, "Hidden" doesn't exist. I believe, it's hidden, if the "prompt" is an empty string.
Maybe it was once valid in 2008. Or it's a copy and paste error from some previous queries I used somewhere else.
Please, be cautious with the properties which I didn't take into the final resultset.

Additionally, I just see: It's not "Datatype" but "Type" to get the parameter datatype.

**Comment by Seddryck** [comment]
Thx for confirmation.
